bin/vcfparser.py

```diff
@@ -0,0 +1,180 @@
+
+    Returns:
+        hgnc_ids     (list): list of hgnc ids in the record
+        alleles      (list): list of alleles in the record
+        consequences (list): list of consequence terms in the record
+        severity     (list): list of consequence term ranks
+    """
+
+    consequences = []
+    hgnc_ids = []
+    severity = []
+    alleles = []
+    mscsq_anno = []
+    for transcript in transcripts:
+        vep_fields = transcript.strip().split("|")
+        csq = vep_fields[csq_ind].split("&")[0]
+        hgnc_id = vep_fields[hgnc_ind]
+        allele = vep_fields[allele_ind].replace("CSQ=", "")
+        consequences.append(csq)
+        hgnc_ids.append(hgnc_id)
+        severity.append(var_csq.index(csq))
+        alleles.append(allele)
+    return hgnc_ids, alleles, consequences, severity
+
+
+def construct_most_severe_consequence_info(line, allele_ind, csq_ind, hgnc_ind, var_csq):
+    """
+    Parse conseqences for each transcript and return HGNC IDs, alleles, and their severity rank
+    based on the term's ranking in the ensembl consequences list.
+
+    Args:
+        line        (str) : Vcf record
+        allele_ind  (int) : Index of the "allele" in the vep annotation record
+        csq_ind     (int) : Index of the "Consequence" in the vep annotation record
+        hgnc_ind    (int) : Index of the "HGNC_ID" in the vep annotation record
+        var_csq     (list): A list of consequence terms ordered by rank
+
+    Returns:
+        columns     (list): A list of fields in the vcf record with most severe consequence added
+                            to the INFO column
+    """
+
+    columns = line.strip().split()
+    transcripts = columns[7].split("CSQ=")[1].split(",")
+    hgnc_ids, alleles, consequences, severity = parse_vep_csq_transcripts(
+        transcripts, allele_ind, csq_ind, hgnc_ind, var_csq
+    )
+    unique_ids = list(set(hgnc_ids))
+    mscsq_anno = []
+    for gene_id in unique_ids:
+        if gene_id != "":
+            indices = find_indices(hgnc_ids, gene_id)
+            alleles_sub = [alleles[i] for i in indices]
+            consequences_sub = [consequences[i] for i in indices]
+            severity_sub = [severity[i] for i in indices]
+            most_severe_csq = consequences_sub[severity_sub.index(min(severity_sub))]
+            most_severe_allele = alleles_sub[severity_sub.index(min(severity_sub))]
+            mscsq_anno.append(gene_id + ":" + most_severe_allele + "|" + most_severe_csq)
+    columns[7] += ";most_severe_consequence=" + ",".join(mscsq_anno)
+    return columns
+
+
+def find_indices(list_to_check, item_to_find):
+    """
+    Get indices of an element in a list
+
+    Args:
+        list_to_check (list)
+        item_to_find  (value)
+
+    Returns:
+        indices       (list)
+    """
+    indices = []
+    for idx, value in enumerate(list_to_check):
+        if value == item_to_find:
+            indices.append(idx)
+    return indices
+
```

Comment:
True! I did think about using numpy, but decided against it as it is not part of the standard python library so it would be an extra module to install and import. And these lists are not that long so the speed advantage that we would get wouldn't be that much either (for a trio, this script take about 2-3 min to finish).